hypothesis-python/src/hypothesis/internal/conjecture/shrinking/length.py

```diff
@@ -52,50 +53,65 @@ def left_is_better(self, left, right):
         return len(left) < len(right)
 
     def run_step(self):
-        # Try to delete as many elements as possible from the sequence, in
-        # (roughly) one pass, from right to left.
-
-        # Starting from the end of the sequence, we try to delete as many
-        # consecutive elements as possible. When we encounter an element that
-        # can't be deleted this way, we skip over it for the rest of the pass,
-        # and continue to its left. This lets us finish a pass in linear time,
-        # but the drawback is that we'll miss some possible deletions of
-        # already-skipped elements.
-        skipped = 0
-
-        # When every element has been deleted or skipped, the pass is complete.
-        while skipped < len(self.current):
-            # Number of remaining elements to the left of the skipped region.
-            # These are all candidates for attempted deletion.
-            candidates = len(self.current) - skipped
-
-            # Take a stable snapshot of the current sequence, so that deleting
-            # elements doesn't mess with our slice indices.
-            start = self.current
-
-            # Delete as many elements as possible (k) from the candidate
-            # region, from right to left. Always retain the skipped elements
-            # at the end. (See diagram below.)
-            find_integer(
-                lambda k: k <= candidates
-                and self.consider(start[: candidates - k] + start[candidates:])
+        # Try to delete as many elements as possible from the sequence, trying
+        # each element no more than once.
+
+        # We convert the sequence to a set of indices. This allows us to more
+        # easily do book-keeping around which elements we've tried removing.
+        initial = self.current
+
+        indices = list(hrange(len(self.current)))
+
+        # The set of indices that we have so far decided to retain.
+        current_subset = set(indices)
+
+        # The set of indices that we still need to consider removing.
+        candidates_for_removal = set(current_subset)
+
+        def consider_set(keep):
+            """Try replacing current_subset with current_subset & keep."""
+
+            keep = keep & current_subset
+            if len(keep) == len(current_subset):
+                return True
+            if self.consider([v for i, v in enumerate(initial) if i in keep]):
+                current_subset.intersection_update(keep)
```

Comment:
Is there a particular reason for using `intersection_update` here, and `&=` elsewhere?

Reply:
It's a scoping thing. `&=` is visually more pleasing but you can't use it for something in an enclosing function scope because you can't assign to an enclosing variable (without the Python 3 only nonlocal annotation)